packages/devcmd/src/process/ProcessExecutor.ts

```diff
@@ -203,29 +329,58 @@ export class ProcessExecutor {
       }),
     ]);
 
-    if (code !== 0) {
+    if (exitCode !== 0) {
+      const nonZeroExitCodeMessage = formatNonZeroExitCodeMessage(processInfo, exitCode);
       const message =
-        `Process '${processInfo.command}' exited with status code ${code}\n\n` +
-        `STDOUT WAS:\n${childStdout}\n\n` +
+        `${nonZeroExitCodeMessage}\n\n` + //
+        `STDOUT WAS:\n${childStdout}\n\n` + //
         `STDERR WAS:\n${childStderr}\n\n`;
-      this.printError(message);
-      throw new Error(message);
-    }
 
-    this.printNotice(`Process ${formatProcessCommand(processInfo)} exited successfully.`);
+      switch (nonZeroExitCodeHandling) {
+        case "printErrorAndThrow":
+          this.printError(message);
+          throw new Error(message);
+        case "printNoticeAndReturn":
+          this.printNotice(message);
+          break;
+        default:
+          throw new Error(`Unknown value for option 'nonZeroExitCodeHandling': '${nonZeroExitCodeHandling}'`);
+      }
+    } else {
+      this.printNotice(`Process ${formatProcessCommand(processInfo)} exited successfully.`);
+    }
 
-    return { stdout: childStdout, stderr: childStderr };
+    return { stdout: childStdout, stderr: childStderr, exitCode, exitSignal };
   }
 }
 
-function childProcessCompletion(childProcess: ChildProcess): Promise<number | null> {
-  return new Promise<number | null>((resolve, reject) => {
+/**
+ * Information about a process's exit status as provided by Node.
+ *
+ * See Node's documentation on the `exit` event of `ChildProcess`:
+ * https://nodejs.org/docs/latest-v12.x/api/child_process.html#child_process_event_exit
+ */
+export interface NodeExitInfo {
+  /**
+   * If the process exited on its own, `exitCode` is the final exit code of the process,
+   * otherwise `null`. Node's API assures that either this or `exitSignal` is always non-`null`.
+   */
+  exitCode: number | null;
+  /**
+   * If the process exited on its own, `exitCode` is the final exit code of the process,
+   * otherwise `null`. Node's API assures that either this or `exitSignal` is always non-`null`.
+   */
+  exitSignal: NodeJS.Signals | null;
```

Comment:
Now that we have introduced the `exitSignal`, maybe we should use it somehow (and not claim that the process failed with `exitCode` `null`). Or why did we introduce it in the first place?

Reply:
Yup, good point. TBH I mostly added it because I finally looked up why `exitCode` can even be null.
I agree that the current error handling for exitCode being null isn't great, but I'd prefer to do that outside of this PR. I'll create a new issue for it.